haystack/preview/document_stores/memory/document_store.py

```diff
@@ -259,3 +271,104 @@ def bm25_retrieval(
             return_document = Document(**doc_fields)
             return_documents.append(return_document)
         return return_documents
+
+    def embedding_retrieval(
+        self,
+        query_embedding: List[float],
+        filters: Optional[Dict[str, Any]] = None,
+        top_k: int = 10,
+        scale_score: bool = True,
+        return_embedding: bool = False,
+    ) -> List[Document]:
+        """
+        Retrieves documents that are most similar to the query embedding using a vector similarity metric.
+
+        :param query_embedding: Embedding of the query.
+        :param filters: A dictionary with filters to narrow down the search space.
+        :param top_k: The number of top documents to retrieve. Default is 10.
+        :param scale_score: Whether to scale the scores of the retrieved Documents. Default is True.
+        :param return_embedding: Whether to return the embedding of the retrieved Documents. Default is False.
+        :return: A list of the top 'k' documents most relevant to the query.
+        """
+        if len(query_embedding) == 0 or not isinstance(query_embedding[0], float):
+            raise ValueError("query_embedding should be a non-empty list of floats.")
+
+        filters = filters or {}
+        all_documents = self.filter_documents(filters=filters)
+
+        documents_with_embeddings = [doc for doc in all_documents if doc.embedding is not None]
+        if len(documents_with_embeddings) == 0:
+            logger.info(
+                "No Documents found with embeddings. Returning empty list. "
+                "To generate embeddings, use a DocumentEmbedder."
+            )
+            return []
+        elif len(documents_with_embeddings) < len(all_documents):
+            logger.info(
+                "Skipping some Documents that don't have an embedding. "
+                "To generate embeddings, use a DocumentEmbedder."
+            )
+
+        # mypy complains that doc.embedding could be None, but we already excluded all Documents with None embedding
+        embedding_sizes = [len(doc.embedding) for doc in documents_with_embeddings]  # type: ignore[arg-type]
```

Comment:
Isn't there a specific torch or numpy error that could be caught instead of iterating over all stored documents (which could make the whole thing slow)?

Reply:
Initially, I had the same idea, but then I found that `numpy` raises a generic `ValueError` with a specific error message ("shapes ... not aligned"), so it's not easy to catch this exact error.

Reply:
I think you can test the error message when you catch an exception. I guess that could save a lot of time! 

(It does depend on the exception though, so let's make sure if it's possible or not)